tests: one test_scalar_prop_validation test case is failing

Hi again,

Here's something else that I noticed while reporting my other one: https://github.com/JPEWdev/shacl2code/issues/5.

One unit test is failing with:

```
FAILED tests/test_python.py::test_scalar_prop_validation[test_class_regex_datetime-value158-ValueError] - Failed: DID NOT RAISE <class 'ValueError'>
```

Potentially relevant information:

```
macOS 12.6.3 (21G419)
python --version
Python 3.8.18
```

Let me know if I could provide any more information or test something out.

----

```
=================================== FAILURES ===================================
__ test_scalar_prop_validation[test_class_regex_datetime-value158-ValueError] __

model = <module 'model' from '/private/var/folders/sy/_sd67yzn5td5flr2_6vf73w40000gp/T/pytest-of-Stanislaw/pytest-0/pythontestcontext0/model.py'>
prop = 'test_class_regex_datetime', value = datetime.datetime(2024, 3, 11, 0, 0)
expect = <class 'ValueError'>

    @pytest.mark.parametrize(
        "prop,value,expect",
        [
            # postive integer
            ("test_class_positive_integer_prop", -1, ValueError),
            ("test_class_positive_integer_prop", 0, ValueError),
            ("test_class_positive_integer_prop", 1, 1),
            ("test_class_positive_integer_prop", False, ValueError),
            ("test_class_positive_integer_prop", True, 1),
            *type_tests("test_class_positive_integer_prop", int),
            # non-negative integer
            ("test_class_nonnegative_integer_prop", -1, ValueError),
            ("test_class_nonnegative_integer_prop", 0, 0),
            ("test_class_nonnegative_integer_prop", 1, 1),
            ("test_class_nonnegative_integer_prop", False, 0),
            ("test_class_nonnegative_integer_prop", True, 1),
            *type_tests("test_class_nonnegative_integer_prop", int),
            # integer
            ("test_class_integer_prop", -1, -1),
            ("test_class_integer_prop", 0, 0),
            ("test_class_integer_prop", 1, 1),
            ("test_class_integer_prop", False, 0),
            ("test_class_integer_prop", True, 1),
            *type_tests("test_class_integer_prop", int),
            # float
            ("test_class_float_prop", -1, -1.0),
            ("test_class_float_prop", -1.0, -1.0),
            ("test_class_float_prop", 0, 0.0),
            ("test_class_float_prop", 0.0, 0.0),
            ("test_class_float_prop", 1, 1.0),
            ("test_class_float_prop", 1.0, 1.0),
            ("test_class_float_prop", False, 0.0),
            ("test_class_float_prop", True, 1.0),
            *type_tests("test_class_float_prop", int, float),
            # boolean
            ("test_class_boolean_prop", True, True),
            ("test_class_boolean_prop", False, False),
            *type_tests("test_class_boolean_prop", bool),
            # datetime
            (
                # Local time
                "test_class_datetime_scalar_prop",
                datetime(2024, 3, 11, 0, 0, 0),
                datetime(2024, 3, 11, 0, 0, 0).astimezone(),
            ),
            (
                # Explict timezone
                "test_class_datetime_scalar_prop",
                datetime(2024, 3, 11, 0, 0, 0, tzinfo=timezone(-timedelta(hours=6))),
                SAME_AS_VALUE,
            ),
            (
                # Explict timezone
                "test_class_datetime_scalar_prop",
                datetime(2024, 3, 11, 0, 0, 0, tzinfo=timezone(timedelta(hours=0))),
                SAME_AS_VALUE,
            ),
            (
                "test_class_datetime_scalar_prop",
                # UTC
                datetime(2024, 3, 11, 0, 0, 0, tzinfo=timezone.utc),
                SAME_AS_VALUE,
            ),
            (
                "test_class_datetime_scalar_prop",
                # Microseconds are ignored
                datetime(2024, 3, 11, 0, 0, 0, 999),
                datetime(2024, 3, 11, 0, 0, 0).astimezone(),
            ),
            (
                "test_class_datetime_scalar_prop",
                # Minutes timezone
                datetime(
                    2024, 3, 11, 0, 0, 0, tzinfo=timezone(timedelta(hours=-1, minutes=21))
                ),
                SAME_AS_VALUE,
            ),
            (
                "test_class_datetime_scalar_prop",
                # Seconds from timezone are dropped
                datetime(
                    2024,
                    3,
                    11,
                    0,
                    0,
                    0,
                    tzinfo=timezone(timedelta(hours=-1, minutes=21, seconds=31)),
                ),
                datetime(
                    2024, 3, 11, 0, 0, 0, tzinfo=timezone(timedelta(hours=-1, minutes=21))
                ),
            ),
            *type_tests("test_class_datetime_scalar_prop", datetime),
            # datetimestamp
            (
                # Local time
                "test_class_datetimestamp_scalar_prop",
                datetime(2024, 3, 11, 0, 0, 0),
                datetime(2024, 3, 11, 0, 0, 0).astimezone(),
            ),
            (
                # Explict timezone
                "test_class_datetimestamp_scalar_prop",
                datetime(2024, 3, 11, 0, 0, 0, tzinfo=timezone(-timedelta(hours=6))),
                SAME_AS_VALUE,
            ),
            (
                # Explict timezone
                "test_class_datetimestamp_scalar_prop",
                datetime(2024, 3, 11, 0, 0, 0, tzinfo=timezone(timedelta(hours=0))),
                SAME_AS_VALUE,
            ),
            (
                "test_class_datetimestamp_scalar_prop",
                # UTC
                datetime(2024, 3, 11, 0, 0, 0, tzinfo=timezone.utc),
                SAME_AS_VALUE,
            ),
            (
                "test_class_datetimestamp_scalar_prop",
                # Microseconds are ignored
                datetime(2024, 3, 11, 0, 0, 0, 999, tzinfo=timezone.utc),
                datetime(2024, 3, 11, 0, 0, 0, tzinfo=timezone.utc),
            ),
            (
                "test_class_datetimestamp_scalar_prop",
                # Minutes timezone
                datetime(
                    2024, 3, 11, 0, 0, 0, tzinfo=timezone(timedelta(hours=-1, minutes=21))
                ),
                SAME_AS_VALUE,
            ),
            (
                "test_class_datetimestamp_scalar_prop",
                # Seconds from timezone are dropped
                datetime(
                    2024,
                    3,
                    11,
                    0,
                    0,
                    0,
                    tzinfo=timezone(timedelta(hours=-1, minutes=21, seconds=31)),
                ),
                datetime(
                    2024, 3, 11, 0, 0, 0, tzinfo=timezone(timedelta(hours=-1, minutes=21))
                ),
            ),
            *type_tests("test_class_datetimestamp_scalar_prop", datetime),
            # String
            ("test_class_string_scalar_prop", "foo", "foo"),
            ("test_class_string_scalar_prop", "", ""),
            *type_tests("test_class_string_scalar_prop", str),
            # Named property
            ("named_property", "foo", "foo"),
            ("named_property", "", ""),
            *type_tests("named_property", str),
            # Enumerated value
            (
                "test_class_enum_prop",
                "http://example.org/enumType/foo",
                "http://example.org/enumType/foo",
            ),
            ("test_class_enum_prop", "foo", ValueError),
            *type_tests("test_class_enum_prop", str),
            # Object
            ("test_class_class_prop", lambda model: model.test_class(), SAME_AS_VALUE),
            (
                "test_class_class_prop",
                lambda model: model.test_derived_class(),
                SAME_AS_VALUE,
            ),
            ("test_class_class_prop", lambda model: model.test_another_class(), TypeError),
            ("test_class_class_prop", lambda model: model.parent_class(), TypeError),
            ("test_class_class_prop", "_:blanknode", "_:blanknode"),
            (
                "test_class_class_prop",
                "http://serialize.example.org/test",
                "http://serialize.example.org/test",
            ),
            *type_tests("test_class_class_prop", str),
            # Pattern validated string
            ("test_class_regex", "foo1", "foo1"),
            ("test_class_regex", "foo2", "foo2"),
            ("test_class_regex", "foo2a", "foo2a"),
            ("test_class_regex", "bar", ValueError),
            ("test_class_regex", "fooa", ValueError),
            ("test_class_regex", "afoo1", ValueError),
            *type_tests("test_class_regex", str),
            # Pattern validated dateTime
            (
                "test_class_regex_datetime",
                datetime(2024, 3, 11, 0, 0, 0, tzinfo=timezone(timedelta(hours=1))),
                SAME_AS_VALUE,
            ),
            (
                "test_class_regex_datetime",
                datetime(2024, 3, 11, 0, 0, 0, tzinfo=timezone(-timedelta(hours=6))),
                ValueError,
            ),
            (
                "test_class_regex_datetime",
                datetime(2024, 3, 11, 0, 0, 0, tzinfo=timezone.utc),
                ValueError,
            ),
            (
                "test_class_regex_datetime",
                datetime(2024, 3, 11, 0, 0, 0),
                ValueError,
            ),
            # Pattern validated dateTimeStamp
            (
                "test_class_regex_datetimestamp",
                datetime(2024, 3, 11, 0, 0, 0, tzinfo=timezone(-timedelta(hours=6))),
                ValueError,
            ),
            (
                "test_class_regex_datetimestamp",
                datetime(2024, 3, 11, 0, 0, 0, tzinfo=timezone(timedelta(hours=0))),
                datetime(2024, 3, 11, 0, 0, 0, tzinfo=timezone.utc),
            ),
            (
                "test_class_regex_datetimestamp",
                datetime(2024, 3, 11, 0, 0, 0, tzinfo=timezone.utc),
                SAME_AS_VALUE,
            ),
            # Property that is a keyword
            ("import_", "foo", "foo"),
            # A property that conflicts with an existing attribute
            ("encode_", "foo", "foo"),
            # Verify that existing attributes cannot be overwritten
            ("encode", "foo", AttributeError),
        ],
    )
    def test_scalar_prop_validation(model, prop, value, expect):
        c = model.test_class()

        if callable(value):
            value = value(model)

        if expect is SAME_AS_VALUE:
            expect = value

        kwargs = {prop: value}

        for cls in model.test_class, model.test_derived_class:
            c = cls()

            if isinstance(expect, type) and issubclass(expect, Exception):
                with pytest.raises(expect):
>                   setattr(c, prop, value)
E                   Failed: DID NOT RAISE <class 'ValueError'>

tests/test_python.py:831: Failed
=============================== warnings summary ===============================
tests/test_model_source.py::test_context_args
tests/test_model_source.py::test_model_errors[bad-reference.jsonld]
tests/test_python.py::test_objset_foreach_type
tests/test_python.py::test_objset_foreach_type
  /Users/Stanislaw/workspace/projects/strictdoc-project/shacl2code/.venv/lib/python3.8/site-packages/pytest_shutil/workspace.py:139: DeprecationWarning: isdir is deprecated; use is_dir
    if self.workspace.isdir():

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform darwin, python 3.8.18-final-0 ----------
Name                                Stmts   Miss  Cover
-------------------------------------------------------
src/shacl2code/__init__.py              4      0   100%
src/shacl2code/__main__.py              4      0   100%
src/shacl2code/context.py             111      1    99%
src/shacl2code/lang/__init__.py         4      0   100%
src/shacl2code/lang/common.py          65      0   100%
src/shacl2code/lang/jinja.py           12      0   100%
src/shacl2code/lang/jsonschema.py      26      0   100%
src/shacl2code/lang/lang.py             8      0   100%
src/shacl2code/lang/python.py          25      0   100%
src/shacl2code/main.py                 68      0   100%
src/shacl2code/model.py               154      0   100%
src/shacl2code/urlcontext.py           13      0   100%
src/shacl2code/version.py               1      0   100%
-------------------------------------------------------
TOTAL                                 495      1    99%

=========================== short test summary info ============================
FAILED tests/test_python.py::test_scalar_prop_validation[test_class_regex_datetime-value158-ValueError] - Failed: DID NOT RAISE <class 'ValueError'>
============ 1 failed, 637 passed, 4 warnings in 283.46s (0:04:43) =============
```

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -3,10 +3,12 @@
 #
 # SPDX-License-Identifier: MIT
 
+import json
+import os
 import pytest
 import shutil
 import subprocess
-import json
+import time
 
 from pytest_server_fixtures.http import SimpleHTTPTestServer
 from pathlib import Path
@@ -61,3 +63,23 @@ def test_jsonschema(model_server, test_context_url):
     )
 
     yield json.loads(p.stdout)
+
+
+@pytest.fixture(scope="function")
+def test_timezone():
+    try:
+        current_tz = os.environ["TZ"]
+    except KeyError:
+        current_tz = None
+
+    os.environ["TZ"] = "TST+02"
+    time.tzset()
+
+    try:
+        yield
+    finally:
+        if current_tz is None:
+            del os.environ["TZ"]
+        else:
+            os.environ["TZ"] = current_tz
+        time.tzset()

--- a/tests/test_python.py
+++ b/tests/test_python.py
@@ -28,6 +28,8 @@ TEST_CONTEXT = THIS_DIR / "data" / "model" / "test-context.json"
 
 SPDX3_CONTEXT_URL = "https://spdx.github.io/spdx-3-model/context.json"
 
+TEST_TZ = timezone(timedelta(hours=-2), name="TST")
+
 
 @pytest.fixture(scope="session")
 def roundtrip(tmp_path_factory, model_server):
@@ -619,8 +621,8 @@ def type_tests(name, *typ):
         (
             # Local time
             "test_class_datetime_scalar_prop",
-            datetime(2024, 3, 11, 0, 0, 0),
-            datetime(2024, 3, 11, 0, 0, 0).astimezone(),
+            lambda _: datetime(2024, 3, 11, 0, 0, 0),
+            datetime(2024, 3, 11, 0, 0, 0, tzinfo=TEST_TZ),
         ),
         (
             # Explict timezone
@@ -643,8 +645,8 @@ def type_tests(name, *typ):
         (
             "test_class_datetime_scalar_prop",
             # Microseconds are ignored
-            datetime(2024, 3, 11, 0, 0, 0, 999),
-            datetime(2024, 3, 11, 0, 0, 0).astimezone(),
+            lambda _: datetime(2024, 3, 11, 0, 0, 0, 999),
+            datetime(2024, 3, 11, 0, 0, 0, tzinfo=TEST_TZ),
         ),
         (
             "test_class_datetime_scalar_prop",
@@ -675,8 +677,8 @@ def type_tests(name, *typ):
         (
             # Local time
             "test_class_datetimestamp_scalar_prop",
-            datetime(2024, 3, 11, 0, 0, 0),
-            datetime(2024, 3, 11, 0, 0, 0).astimezone(),
+            lambda _: datetime(2024, 3, 11, 0, 0, 0),
+            datetime(2024, 3, 11, 0, 0, 0, tzinfo=TEST_TZ),
         ),
         (
             # Explict timezone
@@ -812,7 +814,7 @@ def type_tests(name, *typ):
         ("encode", "foo", AttributeError),
     ],
 )
-def test_scalar_prop_validation(model, prop, value, expect):
+def test_scalar_prop_validation(model, test_timezone, prop, value, expect):
     c = model.test_class()
 
     if callable(value):

